astropy/cosmology/core.py

```diff
@@ -375,11 +388,58 @@ class FlatCosmologyMixin(metaclass=abc.ABCMeta):
     but ``FlatLambdaCDM`` **will** be flat.
     """
 
+    # ===============================================================
+
+    @classmethod  # TODO! make metaclass-method
+    def _get_nonflat_cls(cls: Type[_FCT], kls: Optional[Type[_CT]]=None) -> Type[_CT]:
+        """Find the corresponding non-flat class.
+
+        The class' bases are searched recursively.
+
+        Parameters
+        ----------
+        kls : :class:`astropy.cosmology.Cosmology` class or None, optional
+            If `None` (default) this class is searched.
+
+        Raises
+        ------
+        CosmologyError
+            If more than one non-flat class is found at the same level of the
+            inheritance.
+
+        Returns
+        -------
+        type
+            A Cosmology class that is not a FlatCosmologyMixin subclass.
+        """
+        kls = cls if kls is None else kls  # Which class to search.
+
+        # Find non-flat classes
+        nonflat = [b for b in kls.__bases__ if not issubclass(b, FlatCosmologyMixin)]
+
+        if len(nonflat) > 1:
+            raise CosmologyError("more than one non-flat class in the inheritance")
+        elif not nonflat:  # Recurse higher to find non-flat class
+            for base in kls.__bases__:
+                return cls._get_nonflat_cls(base)
+
+        return nonflat[0]
+
+    _nonflat_cls_ = classproperty(_get_nonflat_cls, lazy=True,
+                                  doc="return the corresponding non-flat class.")
+
+    # ===============================================================
+
     @property
     def is_flat(self):
         """Return `True`, the cosmology is flat."""
         return True
 
+    @property
+    @abc.abstractmethod
+    def equivalent_nonflat(self: _FCT) -> _CT:
+        """Return the equivalent non-flat-class instance of this cosmology."""
```

Comment:
Wouldn't it be better to make this method raise a `NotImplementedError` instead of making it abstract? In that case existing code with user-defined subclasses would not be broken by this update.

Reply:
Hm. That's a possibility. But I don't think it's actually going to be a problem for any users. The `FlatCosmologyMixin` is  like `Cosmology` in no one actually uses it since `FLRW` and `FlatFLRWMixin` are the base classes that 99⨦1% of people will use. In `FlatFLRWMixin` `equivalent_nonflat ` is not abstract. I have yet to see a direct Cosmology or FlatCosmologyMixin subclass in the wild or with a GH search.

Reply:
That motivates why it's not a problem. As to why introducing a new abstract method is desirable... it's so that any hypothetical future cosmology classes that don't inherit from FLRW will have a similar API.

Reply:
I am not a cosmologist myself, so I wouldn't know how `astropy.cosmology` is actually used. I'm going to have to trust your judgment here.